### DISCUSSION TITLE
Unable to deserialize EventHandler class using ObjectMapper

### DISCUSSION THREAD

**Discussion by der4im** [Netflix/conductor#2995] [Q&A] [answered]
Hello,
I am new to conductor and exploring the sample that demonstrate how to use events from the developer labs. I have an event defined in a json file and upon my application startup, I read it from the file and call ObjectMapper to convert that into an EventHandler object. The call to ObjectMapper.readvalue() fails with the following error:

`com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot find a (Map) Key deserializer for type [simple type, class com.google.protobuf.Descriptors$FieldDescriptor]  at [Source: (sun.nio.ch.ChannelInputStream); line: 1, column: 1]
`
Here's a code snippet:

`
try (InputStream i = resource.getInputStream()) {

                EventHandler eventHandlerDef = objectMapper.readValue(i, EventHandler.class);
}
catch (Exception ex) {
     // log exception
}
`

Could someone please help?

**Comment by der4im** [comment]
I also tried to register the events using the /events api. This also fails. Please see error message below. Effectively, both methods of registering events viz. through SDK and through API isn't working for me. I would greatly appreciate if someone can give me some pointers here. 

`{
  "status": 500,
  "message": "INTERNAL_ERROR - JSON parse error: invalid '@type' field when deserializing ProtoBuf Any object; nested exception is com.fasterxml.jackson.databind.JsonMappingException: invalid '@type' field when deserializing ProtoBuf Any object\n at [Source: (PushbackInputStream); line: 15, column: 25] (through reference chain: com.netflix.conductor.common.metadata.events.EventHandler[\"actions\"]->java.util.ArrayList[0]->com.netflix.conductor.common.metadata.events.EventHandler$Action[\"start_workflow\"]->com.netflix.conductor.common.metadata.events.EventHandler$StartWorkflow[\"inputMessage\"])",
  "instance": "b8c69ed6327f",
  "retryable": false
}
`

Here's how I call the api:

`curl -X 'POST' \
  'http://localhost:8080/api/event' \
  -H 'accept: */*' \
  -H 'Content-Type: application/json' \
  -d '{
  "name": "payment_process_event_handler",
  "event": "conductor:store_order_using_event:process_payment_event",
  "condition": "",
  "actions": [
    {
      "action": "start_workflow",
      "start_workflow": {
        "name": "process_payment",
        "version": 1,
        "correlationId": "",
        "input": {
          "sourceWorkflowId": "${workflowInstanceId}"
        },
        "inputMessage": "",
        "taskToDomain": {}
      },
      "complete_task": {},
      "fail_task": {},
      "expandInlineJSON": true
    }
  ],
  "active": true,
  "evaluatorType": ""
}'`

**Comment by apanicker-nflx** [comment, marked as answer]
Can you remove the `"inputMessage": "",` from your post data and give it a try?

**Reply by der4im** [reply]
Thanks for your response. Yes, I had tried that earlier and that had worked. However, I would like to register the event handlers using the SDK. How can I get that working? I have my event handler spec in a json file. I would like to read them, convert them into EventHandler objects and then call the SDK method:

`EventClient.registerEventHandler(EventHandler eventHandler)` 

How can I do this? The call to objectMapper.readValue() is failing.

**Reply by der4im** [reply]
Could someone please help?

**Reply by apanicker-nflx** [reply]
@der4im How are you initializing the ObjectMapper that you are using to register the event handler?
I would recommend using the ObjectMapper that is provided by Conductor for this purpose. Please refer [this](https://github.com/Netflix/conductor/blob/main/common/src/main/java/com/netflix/conductor/common/config/ObjectMapperProvider.java#L40)

**Reply by der4im** [reply]
I mentioned this in my initial post but here it is again:
`try (InputStream i = resource.getInputStream()) {

            EventHandler eventHandlerDef = objectMapper.readValue(i, EventHandler.class);

}
catch (Exception ex) {
// log exception
}`

**Reply by jxu-nflx** [reply]
@der4im where does your objectMapper come from? Did you apply the same config as [this](https://github.com/Netflix/conductor/blob/main/common/src/main/java/com/netflix/conductor/common/config/ObjectMapperProvider.java#L40)
```
objectMapper.registerModule(new JsonProtoModule());
```

**Reply by der4im** [reply]
No, it is the default Java ObjectMapper.

**Reply by der4im** [reply]
Using the netflix objectmapper worked. Thank you folks for helping!
